nixos/virtualisation.podman: Init module

adisbladis opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This is another stab at making a module for `podman`, a drop-in replacement for Docker.

Unlike previous attempts ( https://github.com/NixOS/nixpkgs/pull/68343 & https://github.com/NixOS/nixpkgs/pull/54925 ) this tries to stay very close to upstream defaults and introduce minimal configuration.

I also have tests and release notes. 
I have been running a very similar module in my own configuration for quite some time without issues.
It works just fine rootless too with the following configuration in `configuration.nix`:
``` nix
virtualisation.podman.enable = true;
virtualisation.podman.dockerCompat = true;
virtualisation.containers.users = [ "adisbladis" "aoeu" ];
```
~Should we add a config option to set up subuid/subgid ranges for users in the podman module maybe?~ - done

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @vdemeester @peterhoeg @nlewo 

Closes https://github.com/NixOS/nixpkgs/issues/65202

adisbladis committed:
nixos/virtualisation.podman: Init module

peterhoeg commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,127 @@
+
+  config = lib.mkIf cfg.enable {
+
+    environment.systemPackages = [
```
Instead of sticking all these into the global env, why not generate `libpod.conf` instead which points to the right binaries?

adisbladis replied:
`/etc/containers/libpod.conf` seems not to be used when I run containers as my normal user. Unless I stick `conmon` & friends in the global env I'd have to replicate `libpod.conf` in my home directory.

peterhoeg commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,127 @@
+
+    environment.systemPackages = [
+      pkgs.podman # Docker compat
+      pkgs.runc # Container runtime
```
I used `crun` instead of `runc` for rootless support as `runc` at least didn't use to support it, but maybe that has changed. This is likely something you want to make configurable.

vdemeester replied:
Hum I am using `runc` with rootless support. I think in the latests rc(s) it supports it

adisbladis replied:
`runc` works just fine rootless.

peterhoeg commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,127 @@
+
+    environment.etc."containers/policy.json".source =
+      if cfg.policy != {} then pkgs.writeText "policy.json" (builtins.toJSON cfg.policy)
+      else "${pkgs.skopeo.src}/default-policy.json";
```
Excellent idea to pull it in like this! :+1:

peterhoeg commented:
This is nice and much nicer than mine (a lot less code which is great!).

> Should we add a config option to set up subuid/subgid ranges for users in the podman module maybe?

Yes please.

adisbladis commented:
> Should we add a config option to set up subuid/subgid ranges for users in the podman module maybe?
> Yes please.

Can you think of a nice interface for this?

adisbladis commented:
I've opted to make subuid/subgid users a simple list:
```
    virtualisation.podman.users = [
      "adisbladis"
    ];
```

zowoq commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,174 @@
+
+    environment.systemPackages = [
+      pkgs.podman # Docker compat
+      pkgs.runc # Default container runtime
```
With cgroups v2 podman defaults to `crun` and will error if it isn't available, even if `runc` is available.

adisbladis replied:
Also added crun!

zowoq commented:
`podman` and related packages have a [team](https://github.com/NixOS/nixpkgs/blob/master/maintainers/team-list.nix) if you want to add it as a a maintainer for the module.

cc @saschagrunert

adisbladis commented:
I've also moved declarative docker containers to podman in https://github.com/adisbladis/nixpkgs/tree/podman-declarative-containers. This will be a follow-up PR.

adisbladis committed:
nixos/podman: Add maintainer team & add myself to podman team

adisbladis commented:
> podman and related packages have a team if you want to add it as a maintainer for the module.

Done! also added myself to the podman team.

adisbladis commented:
@GrahamcOfBorg eval
@GrahamcOfBorg test podman

adisbladis commented:
@GrahamcOfBorg eval
@GrahamcOfBorg test podman

adisbladis commented:
Unless I get some very strong objections my plan is to move fast on this and merge soon.

buckley310 commented:
Is it a good idea to map all podman users to the same 100000+65536 range?

zowoq commented on maintainers/team-list.nix:
```diff
@@ -33,6 +33,7 @@ with lib.maintainers; {
 
   podman = {
```
Could you update the team scope please?
```
"Maintain Podman related packages and modules."
```

zowoq commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,186 @@
+      pkgs.runc # Default container runtime
+      pkgs.crun # Default container runtime (cgroups v2)
+      pkgs.conmon # Container runtime monitor
+      pkgs.skopeo # Interact with container registry
```
Not sure if `buildah` should here or not, people might expect it to be shipped with `podman`/`skopeo`?

Don't feel strongly either way.

adisbladis replied:
You can still call `podman build` without explicitly adding `buildah`. My feeling is that it doesn't really belong in this module but I could easily be convinced otherwise.

zowoq commented on nixos/tests/podman.nix:
```diff
@@ -0,0 +1,30 @@
+import ./make-test-python.nix (
+  { pkgs, ... }: {
+    name = "podman";
+    meta = with pkgs.stdenv.lib.maintainers; {
+      maintainers = [ adisbladis ];
```
Could use the team here as well?

```suggestion
    meta = with pkgs.stdenv.lib {
      maintainers = teams.podman.members;;
```

zowoq commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,186 @@
+
+    environment.etc."containers/libpod.conf".text = ''
+      cni_plugin_dir = ["${pkgs.cni-plugins}/bin/"]
+      cni_config_dir = "/etc/cni/net.d/"
```
IIRC `cni-plugins`, etc. doesn't work with rootless so could make this optional, might not be worth the effort though.

adisbladis replied:
`/etc/libpod.conf` is not loaded when running rootless anyway.

zowoq replied:
Yeah, so pulling in the package and generating the config is only needed for non-rootless?

zowoq reviewed (commented):
Thanks for working on this!

Tested, works well. Looks great, just a couple of nits.

adisbladis commented:
> Is it a good idea to map all podman users to the same 100000+65536 range?

No that was an oversight. I've fixed this.

saschagrunert commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,194 @@
+      pkgs.runc # Default container runtime
+      pkgs.crun # Default container runtime (cgroups v2)
+      pkgs.conmon # Container runtime monitor
+      pkgs.skopeo # Interact with container registry
```
What is the reason we ship skopeo with podman?

adisbladis replied:
I think it's just a remnant from my personal module. We can remove Skopeo as podman also knows how to interact with registries.

saschagrunert commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -0,0 +1,194 @@
+      pkgs.fuse-overlayfs # CoW for images, much faster than default vfs
+      pkgs.utillinux # nsenter
+      pkgs.iptables
+    ]
```
Do we have to add `cni-plugins` here as well?

adisbladis replied:
I did not have to add it, but it's a good idea.

adisbladis committed:
nixos.virtualisation.containers: Init common /etc/containers configuration module

What's happening now is that both cri-o and podman are creating
/etc/containers/policy.json.

By splitting out the creation of configuration files we can make the
podman module leaner & compose better with other container software.

adisbladis commented:
I have created a new `virtualisation.containers` module. 
The `cri-o` module was also creating `/etc/containers/policy.json` so these modules would be mutually exclusive.
With this change things composes much better with multiple container runtimes.
This leads to a much smaller `podman` module with a better defined scope.

adisbladis committed:
nixos.virtualisation: Move containers.nix to nixos-containers.nix

In anticipation of the new containers module.

adisbladis commented:
@GrahamcOfBorg test podman

talyz reviewed (approved):
Looks really good! The only issue I experienced was this message:
```
ERRO[0520] unable to close namespace: "close /proc/335/ns/user: bad file descriptor"
```
when exiting a container, seemingly at random. This might just be a podman bug and it doesn't seem to cause any problems.

adisbladis commented:
@talyz That looks to be https://github.com/containers/libpod/issues/5873.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/podman-and-docker-fail-to-mount/11709/3
